### PR TITLE
[Extra/rest 2.4] add gas used

### PR DESCRIPTION
**Pull request by satawatnack**
## Implementation details
- add prepare gas used and execute gas used on emitter and flusher

> Note: 
> ```
> ALTER TABLE public.requests ADD prepare_gas_used integer not null default 0;
> ALTER TABLE public.requests ADD execute_gas_used integer not null default 0;
> ```

## Please ensure the following requirements are met before submitting a pull request:

- [x] The pull request is targeted against the correct target branch
- [ ] The pull request is linked to an issue with appropriate discussion and an accepted design OR is linked to a spec that describes the work.
- [x] The pull request includes a description of the implementation/work done in detail.
- [ ] The pull request includes any and all appropriate unit/integration tests
- [ ] You have added a relevant changelog entry to `CHANGELOG_UNRELEASED.md`
- [x] You have re-reviewed the files affected by the pull request (e.g. using the `Files changed` tab in the Github PR explorer)

### CODE REVIEW

File: hooks/emitter/channel.go [hunk 1]
```diff
@@ -181,6 +182,11 @@ func (h *Hook) extractOracleRequestPacket(
 	err := oracletypes.ModuleCdc.UnmarshalJSON(dataOfPacket, &data)
 	if err == nil {
 		if events, ok := evMap[oracletypes.EventTypeRequest+"."+oracletypes.AttributeKeyID]; ok {
+			var prepareGasUsedI interface{}
+			if prepareGasUsed, ok := evMap[oracletypes.EventTypeRequest+"."+oracletypes.AttributeKeyGasUsed]; ok {
+				prepareGasUsedI = oraclekeeper.ConvertToGas(common.Atoui(prepareGasUsed[0]))
```
Comment (colmazia) [inline hooks/emitter/channel.go:187]: why interface{} na, can't we define a struct?

File: hooks/emitter/oracle.go [hunk 2]
```diff
@@ -104,24 +106,32 @@ func (app *Hook) emitReportAndRawReport(
 	}
 }
 
-func (h *Hook) emitUpdateResult(ctx sdk.Context, id types.RequestID, reason string) {
+func (h *Hook) emitUpdateResult(ctx sdk.Context, id types.RequestID, gasUsedI interface{}, reason string) {
```
Comment (RogerKSI) [inline hooks/emitter/oracle.go:109]: should we just call it executeGasUsed?  and use any instead of interface{}

File: hooks/emitter/oracle.go [hunk 3]
```diff
@@ -208,14 +220,20 @@ func (h *Hook) handleMsgEditOracleScript(
 
 // handleEventRequestExecute implements emitter handler for EventRequestExecute.
 func (h *Hook) handleEventRequestExecute(ctx sdk.Context, evMap common.EvMap) {
+	var gasUsedI interface{}
+	if gasUsedMap, ok := evMap[types.EventTypeResolve+"."+types.AttributeKeyGasUsed]; ok {
```
Comment (RogerKSI) [inline hooks/emitter/oracle.go:224]: the name `map` doesn't align with others. 

Should we write helper functions such as getPrepareGasUsed(evMap), getExecuteGasUsed(evMap)? What do you think?

File: hooks/emitter/oracle.go [hunk 4]
```diff
@@ -4,7 +4,9 @@ import (
 	sdk "github.com/cosmos/cosmos-sdk/types"
 
 	"github.com/bandprotocol/chain/v2/hooks/common"
+	oraclekeeper "github.com/bandprotocol/chain/v2/x/oracle/keeper"
 	"github.com/bandprotocol/chain/v2/x/oracle/types"
+	oracletypes "github.com/bandprotocol/chain/v2/x/oracle/types"
```
Comment (RogerKSI) [inline hooks/emitter/oracle.go:9]: duplicate import on oracletypes

Comment (RogerKSI) [inline hooks/emitter/oracle.go:9]: duplicate import on oracletypes

**Review by RogerKSI — CHANGES_REQUESTED** [review]

File: hooks/emitter/oracle.go (hunk 3 above)
Comment (satawatnack) [reply hooks/emitter/oracle.go:224]: I think no need to have helper functions.

File: flusher/flusher/db.py [hunk 5]
```diff
@@ -188,7 +188,9 @@ def Column(*args, **kwargs):
     Column("min_count", sa.Integer),
```
Comment (RogerKSI) [inline flusher/flusher/db.py:188]: please update hasura table as well.

File: hooks/emitter/channel.go [hunk 6]
```diff
@@ -181,6 +182,11 @@ func (h *Hook) extractOracleRequestPacket(
 	err := oracletypes.ModuleCdc.UnmarshalJSON(dataOfPacket, &data)
 	if err == nil {
 		if events, ok := evMap[oracletypes.EventTypeRequest+"."+oracletypes.AttributeKeyID]; ok {
+			var prepareGasUsedI interface{}
+			if prepareGasUsed, ok := evMap[oracletypes.EventTypeRequest+"."+oracletypes.AttributeKeyGasUsed]; ok {
```
Comment (RogerKSI) [inline hooks/emitter/channel.go:186]: please align name with other part.

File: x/oracle/keeper/owasm.go [hunk 7]
```diff
@@ -20,6 +20,10 @@ func ConvertToOwasmGas(cosmos uint64) uint64 {
 	return uint64(cosmos * gasConversionFactor)
 }
 
+func ConvertToGas(owasm uint64) uint64 {
+	return uint64(owasm / gasConversionFactor)
```
Comment (taobun) [inline x/oracle/keeper/owasm.go:24]: Should round up?

File: flusher/flusher/db.py [hunk 8]
```diff
@@ -188,7 +188,9 @@ def Column(*args, **kwargs):
...
     Column("fee_limit", sa.String),
     Column("prepare_gas", sa.Integer),
+    Column("prepare_gas_used", sa.Integer, nullable=True),
```
Comment (taobun) [inline flusher/flusher/db.py:191]: Use 0 as default should be ok instead of null

**Review by taobun — CHANGES_REQUESTED** [review]

**Review by RogerKSI — APPROVED** [review]

**Review by taobun — APPROVED** [review]

**Review by colmazia — APPROVED** [review]

**Comment by RogerKSI** [comment]
@babybunn This one is live on testnet na krub.

**Pull request merged**